planner/core/rule_count_star_rewriter.go

```diff
@@ -0,0 +1,100 @@
+package core
+
+import (
+	"context"
+
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/expression/aggregation"
+	"github.com/pingcap/tidb/parser/mysql"
+)
+
+/**
+The countStarRewriter is used to rewrite count(*) -> count(not null column)
+
+Pattern:
+LogcialAggregation
+       |
+   DataSource
+
+Optimize:
+Table
+  <k1 bool not null, k2 int null, k3 bigint not null>
+
+Case1 there are columns exists in datasource
+Query: select count(*) from table where k3=1
+CountStarRewriterRule: pick the shortest not null column exists in datasource
+Rewritten Query: select count(k3) from table where k3=1
+
+Case2 there is no columns exists in datasource
+Query: select count(*) from table
+ColumnPruningRule: pick k1 as the shortest not null column from origin table @Function.preferNotNullColumnFromTable
+                   datasource.columns: k1
+CountStarRewriterRule: rewrite count(*) -> count(k1)
+Rewritten Query: select count(k1) from table
+
+Attention:
+Since count(*) is directly translated into count(1) during grammar parsing,
+the rewritten pattern actually matches count(1)
+*/
+
+type countStarRewriter struct {
+}
+
+func (c *countStarRewriter) optimize(ctx context.Context, p LogicalPlan, opt *logicalOptimizeOp) (LogicalPlan, error) {
+	return c.countStarRewriter(p, opt)
+}
+
+func (c *countStarRewriter) countStarRewriter(p LogicalPlan, opt *logicalOptimizeOp) (LogicalPlan, error) {
+	// match pattern agg(count(*)) -> datasource
+	if agg, ok := p.(*LogicalAggregation); ok {
+		// todo if all kinds of data source could be matched ?
+		if dataSource, ok := agg.Children()[0].(*DataSource); ok {
+			for _, aggFunc := range agg.AggFuncs {
+				if aggFunc.Name == "count" && len(aggFunc.Args) == 1 {
```

Comment:
select count(distinct 1) from foo
we should not rewrite this query, right? But seems like we may rewrite it to `select count(distinct pk_col) from foo`.

Reply:
This optimization also applies on `select count(*), count(a), count(b) from table`. Can we have a test case to cover this?

Reply:
It looks like it also support group by aggregation. Can you also add a test for group by query?
`select c1, count(*) from tab group by c1`?

Reply:
The count(distinct) and group by shouldn't be written. So I add the unit test for them.
Also it includes `count(*), count(a)`